File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        left = 0
        n = len(height)
        right = n - 1
        ans = 0
        while left < right:
            ans = max(ans, (right - left) * min(height[right], height[left]))
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
                            
        return ans

File: 0011-container-with-most-water/0011-container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        // brute force: O(n^2)
        // check each pair
        
        // two pointers: O(n)
        int maxVolume = 0;
        int left = 0;
        int right = height.size() - 1;
        while (left < right) {           
            maxVolume = max(maxVolume, (right - left) * min(height[left], height[right]));
            if (height[left] > height[right])
                right--;
            else
                left++;
        }
        
        return maxVolume;
    }
};